File: 3.1/A1042.cpp
/**
 * 注意序号和纸牌号的关系
 * 另有该题的java实现版，更直接
 */


#include <cstdio>
const int N = 54;
char mp[5] = {'S','H','C','D','J'};
int start[N+1], end[N+1], order[N+1];
int main(){
	int K;
	scanf("%d",&K);
	for (int i = 1; i <= N; i++){
		start[i] = i;
	}
	for (int i = 1; i <= N; i++){
		scanf("%d",&order[i]);
	}
	for (int i = 0; i < K; i++){
		for (int j = 1; j <= N; j++){
			end[order[j]] = start[j];
		}
		for (int i = 1; i <= N; i++){
			start[i] = end[i];
		}
	}
	for (int i = 1; i <= N-1; i++){
		end[i]--;
		printf ("%c%d ", mp[end[i] / 13], end[i] % 13 + 1);
	}
	printf ("%c%d",mp[end[54] / 13],end[54] % 13 + 1);
}

File: 3.1/A1042.java
import java.util.Arrays;
import java.util.Scanner;

public class A1042 {	//在pat中要用 Main


    public static void main(String[] args) {
        String[] startArray = new String[55];
        String[] end = new String[55];
        int[] order = new int[55];

        for (int i = 1; i <= 13; i++){
            startArray[i] = "S" + i;
        }
        for (int i = 1; i <= 13; i++){
            startArray[i + 13] = "H" + i;
        }
        for (int i = 1; i <= 13; i++){
            startArray[i + 26] = "C" + i;
        }
        for (int i = 1; i <= 13; i++){
            startArray[i + 39] = "D" + i;
        }
        startArray[53] = "J1";
        startArray[54] = "J2";


        Scanner in = new Scanner(System.in);
        int K;
        K = in.nextInt();
        for (int i = 1; i <= 54; i++){
            order[i] = in.nextInt();
        }
        for (int i = 0; i < K; i++){
            for (int j = 1; j <= 54; j++){
                end[order[j]] = startArray[j];
            }
            for (int j = 1; j <= 54; j++){
                startArray[j] = end[j];
            }
        }
        for (int i = 1; i <= 53; i++){
            System.out.print(end[i]+" ");
        }
        System.out.println(end[54]);

    }

}
